packages/contractkit/src/wallets/signers/ledger-signer.ts

```diff
@@ -0,0 +1,90 @@
+import { trimLeading0x } from '@celo/utils/lib/address'
+import { EIP712TypedData, generateTypedDataHash } from '../../utils/sign-typed-data-utils'
+import { RLPEncodedTx } from '../../utils/signing-utils'
+import { Signer } from './signer'
+import { TransportStatusError } from '@ledgerhq/errors'
+import { transportErrorFriendlyMessage } from '../../utils/ledger-utils'
+import * as ethUtil from 'ethereumjs-util'
+
+/**
+ * Signs the EVM transaction with a Ledger device
+ */
+export class LedgerSigner implements Signer {
+  private ledger: any
+  private derivationPath: string
+
+  constructor(ledger: any, derivationPath: string) {
+    this.ledger = ledger
+    this.derivationPath = derivationPath
+  }
+
+  getNativeKey(): string {
+    return this.derivationPath
+  }
+
+  async signTransaction(
+    addToV: number,
+    encodedTx: RLPEncodedTx
+  ): Promise<{ v: string; r: string; s: string }> {
+    try {
+      const signature = await this.ledger!.signTransaction(
+        this.derivationPath,
+        trimLeading0x(encodedTx.rlpEncode) // the ledger requires the rlpEncode without the leading 0x
+      )
+      // EIP155 support. check/recalc signature v value.
+      const rv = parseInt(signature.v, 16)
+      // tslint:disable-next-line: no-bitwise
+      if (rv !== addToV && (rv & addToV) !== rv) {
+        addToV += 1 // add signature v bit.
+      }
+      signature.v = addToV.toString(16)
+      return signature
+    } catch (error) {
+      if (error instanceof TransportStatusError) {
+        transportErrorFriendlyMessage(error)
+      }
+      throw error
+    }
+  }
+
+  async signPersonalMessage(
+    data: string
+  ): Promise<{ v: number; r: Buffer | Uint8Array; s: Buffer | Uint8Array }> {
+    try {
+      const dataBuff = ethUtil.toBuffer(data)
+      const msgHashBuff = ethUtil.hashPersonalMessage(dataBuff)
+      const trimmedMsgHash = trimLeading0x(msgHashBuff.toString('hex'))
```

Comment:
@gastonponti previously this was `trimLeading0x(data)` which I believe is incorrect. We also weren't calling `ethUtil.hashPersonalMessage(dataBuff)` before. Check out the ledger unit tests for example:
https://github.com/LedgerHQ/ledgerjs/blob/ae32cfff7930892c32e38f9c74ac6b4fa193b19b/packages/hw-app-eth/tests/Eth.test.js#L68

Reply:
The message is first hashed in [Accounts.ts](https://github.com/celo-org/celo-monorepo/blob/16622e258bf68c4d56ffb347a3bc88d4daaee9ce/packages/contractkit/src/wrappers/Accounts.ts#L399)
 `const hash = Web3.utils.soliditySha3({ type: 'address', value: address })`
It then gets routed through web3->CeloProvider where it ends up calling signPersonalMessage on the LedgerWallet.
This then calls [ethUtil.hashPersonalMessage](https://github.com/ethereumjs/ethereumjs-util/blob/ae4c3009edd8307b1ef3bebc2dc03c9c2fdf1b40/src/signature.ts#L136) which formats it as an Ethereum message to be signed. This is consistent with the in-memory signer which performs the [hashPersonalMessage](https://github.com/celo-org/celo-monorepo/blob/8b54e4803a67eeb9ac772e8b9e261446623ede41/packages/contractkit/src/wallets/default-wallet.ts#L62) operation.

Once it's signed, `getParsedSignatureOfAddress()` calls `parseSignature(hash, signature, signer)` which takes the original hash, adds the Ethereum prefix (that you pointed out above), and performs ecrecover to verify the signature. Since the signed message and this validating message both have the prefix, the original signer can be recovered and the validation step should return true.

Given that ledger wallet was not adding the prefix previously, I don't know how the verification was working. I'll test it out now using a ledger device.

Reply:
After more digging with @gastonponti, we found that:
1. The ledger signer app adds the "Ethereum" prefix to all `signPersonalMessage()` calls
2. [EIP712 expects this prefix for signed messages](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md#eth_signtypeddata)

This means that the ledger signer is correct, but the in-memory signer had a bug where it left out the "Ethereum" prefix. That's been fixed in this PR.